src/V3Const.cpp

```diff
@@ -76,6 +76,418 @@ class ConstVarFindVisitor final : public AstNVisitor {
+                    // ^{x[0], ~x[0], x[2], x[3]} === ~^{x[2], x[3]}
+                    UASSERT_OBJ(sameFlag, m_refp, "Only true is set in Xor tree");
+                    m_bitPolarity.setBit(bit, 'x');
+                } else {  // And, Or
+                    // Can ignore this nodep as the bit is already registered
+                    if (sameFlag) return;  // a & a == a, b | b == b
+                    // Otherwise result is constant
+                    m_constResult = m_parent->isAndTree() ? 0 : 1;
+                    m_bitPolarity.setAllBitsX();  // The variable is not referred anymore
+                }
+            }
+        }
+        AstNode* getResult() const {
+            FileLine* fl = m_refp->fileline();
+            AstNode* srcp = VN_CAST(m_refp->backp(), WordSel);
+            if (!srcp) srcp = m_refp;
+            const int width = widthOfRef(m_refp);
+
+            if (hasConstantResult())
+                return new AstConst{fl,
+                                    V3Number{srcp, width, static_cast<vluint32_t>(m_constResult)}};
+
+            AstConst* maskValuep = new AstConst{fl, V3Number{srcp, width, 0}};
+            maskValuep->num().opBitsNonX(m_bitPolarity);  // 'x' -> 0, 0->1, 1->1
+            // Let AstConst be in lhs as it is the common convention
+            AstAnd* maskedp = new AstAnd{fl, maskValuep, srcp->cloneTree(false)};
+            AstNode* resultp;
+            if (m_parent->isXorTree()) {
+                resultp = new AstRedXor{fl, maskedp};
+                resultp->dtypep()->widthForce(width, 1);
+            } else {
+                AstConst* compValuep = maskValuep->cloneTree(false);
+                compValuep->num().opBitsOne(m_bitPolarity);  // 'x'->0, 0->0, 1->1
+                if (m_parent->isAndTree()) {
+                    resultp = new AstEq{fl, compValuep, maskedp};
+                } else {  // Or
+                    compValuep->num().opXor(V3Number{compValuep->num()}, maskValuep->num());
+                    resultp = new AstNeq{fl, compValuep, maskedp};
+                }
+            }
+            return resultp;
+        }
+
+        // CONSTRUCTORS
+        VarInfo(ConstBitOpTreeVisitor* parent, AstVarRef* refp)
+            : m_parent(parent)
+            , m_refp(refp)
+            , m_bitPolarity(refp, widthOfRef(refp)) {
+            m_bitPolarity.setAllBitsX();
+        }
+    };
+
+    // MEMBERS
+    bool m_failed = false;
+    bool m_polarity = true;  // Flip when AstNot comes
+    int m_ops = 0;  // Number of operations such as And, Or, Xor, Sel...
+    int m_lsb = 0;  // Current LSB
+    LeafInfo* m_leafp = nullptr;  // AstConst or AstVarRef that currently looking for
+    AstNode* m_rootp;  // Root of this AST subtree
+    AstNode* m_curOpp = nullptr;  // The node that should be added to m_frozenNodes
+
+    AstUser4InUse m_inuser4;
+    std::vector<AstNode*> m_frozenNodes;  // Nodes that cannot be optimized
+    std::vector<VarInfo*> m_varInfo;
```

Comment:
```suggestion
    std::vector<VarInfo*> m_varInfos;
```

Reply:
Done in acef8e00934ba1dcc1510102282071d94118de35